plaidml/op/lib/ops.cc

```diff
@@ -466,6 +467,42 @@ Value binary_crossentropy(const Value& value) {
   return Value{OverrideGrads(deriv, std::vector<Tensor>{T, P}, O)};
 }
 
+Value broadcast(const Value& value) {
+  IVLOG(1, "broadcast");
+  auto args = value.as_tuple();
+  if (args.size() != 3) {
+    throw std::runtime_error(llvm::formatv("PlaidML broadcast op expects 3 arguments (received {0})", args.size()));
+  }
+
+  auto I = args[0].as_tensor();
+  auto input_shape = I.compute_shape().sizes();
```

Comment:
The compile-time perf impact of computing shape each time we broadcast is probably going to be pretty bad. Do we care about that currently in the context of this commit?

If so, switching to an `expand_dims` + autobroadcast implementation should be able to avoid this. The costs of that would be reimplementing the entire thing & the new version would have many more contractions (which should all be optimized away, as they're just reinterpreting the same underlying data with a new shape; but I'm not sure whether we have the relevant optimizations implemented yet).

Reply:
I want to hear what Frank says, but here's my perspective:
The `broadcast` op is primarily to be used in TensorFlow for now, and perhaps in other framework implementations later. Right now, our main focus is getting a few networks up and running, with performance being a secondary priority. It's important to note performance hiccups like this, but I don't think it's our top priority to fix them now.

Reply:
OK, `compute_shape` indeed should be avoided, however, how could we make this work without it? I have a feeling we need a new primitive. For now I'm fine with letting it slide and dealing with this later. But I do wonder how we'd make this work if we wanted to support 'deferred' shapes and it so happens that the `?` dim resolves to `1`.

Reply:
My idea was the following:
Use `expand_dims` to add size 1 dimensions where necessary (to the front of the lower-dimensional tensor in Numpy-style, to the back of the lower-dimensional tensor in that style, or where explicitly requested in the explicit style). Now the tensors are the same rank, and we can rely on autobroadcasting, since when all tensors have the same number of dimensions there is only one legal broadcast so it doesn't matter what the broadcast style is.

Reply:
But it sounds like there's consensus around deferring this to later.

Reply:
The exception to "only one legal broadcast" is if you're not broadcasting 2 tensors to their minimal shared shape, but instead explicitly "upbroadcasting" a tensor to a specific size, e.g. explicitly broadcasting a `(1, 3)` tensor to shape `(4, 3)` (or e.g. a `(?, 3)` tensor where `?` = `1` or `4` to `(4, 3)`); we _could_ do this with an autobroadcast technique by e.g. adding `zeroes_like(target_shape)` to the tensor with the expectation that optimization passes will remove everything but the broadcast. Not sure if that's a good idea.

Reply:
Oh I see. Yes I was wondering what the `broadcast_axes` parameter was for, but it seems like we could just have explicit mode for now and then allow users to specify where the `1`s should be inserted.